zcash_proofs/src/lib.rs

```diff
@@ -206,3 +431,81 @@ pub fn parse_parameters<R: io::Read>(
         sprout_vk,
     }
 }
+
+/// Check if the size of the file at `params_path` matches `expected_bytes`,
+/// using filesystem metadata.
+///
+/// Returns an error containing `name` and `params_source` on failure.
+fn verify_file_size(
+    params_path: &Path,
+    expected_bytes: u64,
+    name: &str,
+    params_source: &str,
+) -> Result<(), io::Error> {
+    let file_size = std::fs::metadata(&params_path)?.len();
+
+    if file_size != expected_bytes {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!(
+                "{} failed validation:\n\
+                 expected: {} bytes,\n\
+                 actual:   {} bytes from {:?}",
+                name, expected_bytes, file_size, params_source,
+            ),
+        ));
+    }
+
+    Ok(())
+}
+
+/// Check if the Blake2b hash from `hash_reader` matches `expected_hash`,
+/// while streaming from `data` into `sink`.
```

Comment:
What `data` is this comment referring to?

Reply:
I think the comment is outdated, the `data` is inside `hash_reader`.

I'll rebase this PR and fix up the comment.

Reply:
```suggestion
/// while streaming from `hash_reader` into `sink`.
```